Modules/_ctypes/_ctypes.c

```diff
@@ -3774,19 +3794,52 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
         return NULL;
     }
 #else
+    #ifdef __CYGWIN__
+        //dlerror() isn't very helpful on cygwin */
+    #else
+        #define USE_DLERROR
+        /* dlerror() always returns the latest error.
+         *
+         * Clear the previous value before calling dlsym(),
+         * to ensure we can tell if our call resulted in an error.
+         */
+        (void)dlerror();
+    #endif
     address = (PPROC)dlsym(handle, name);
-    if (!address) {
-#ifdef __CYGWIN__
-/* dlerror() isn't very helpful on cygwin */
+
+    if (address) {
+        goto dlsym_ok;
+    }
+    else {
+#ifdef USE_DLERROR
+        // This assumes the error message is UTF-8 (or ASCII).
+        // Investigate if this can cause problems.
+        const char *dlerr = dlerror();
+        if (dlerr) {
+            PyObject *message = PyUnicode_DecodeLocale(dlerr, "strict");
+            if (message) {
+                PyErr_SetObject(PyExc_AttributeError, message);
+                Py_DECREF(ftuple);
+                return NULL;
+            }
+            // Ignore errors from converting the message to str
+            PyErr_Clear();
+        }
+#endif
+#undef USE_DLERROR
```

Comment:
nitpicks to make this easier to read:
- Consider moving this `#undef` after the final `#endif`, so the `#undef USE_DLERROR`...`#undef USE_DLERROR` region of code doesn't cross `{`...`}` and `#if`...`#endif` blocks
- Consider indenting `#ifdef USE_DLERROR` & its `#endif`.